### ISSUE TITLE
a question about function"eval()"

### ISSUE THREAD

**Issue by wowotou1022** [ando-khachatryan/HiDDeN#13]
Hello, I would like to ask a question. Why does adding function "eval()" in the validation mode raise "bitwise-error"? After removing "eval()", the "bitwise-error" of the validation mode is similar to the training mode's.
looking forward to your reply.

**Comment by ando-khachatryan** [comment]
eval() switches the network to evaluation mode, which changes the behavior of BatchNorm. See for instance these: 
https://pytorch.org/docs/stable/nn.html#torch.nn.BatchNorm2d
https://github.com/pytorch/pytorch/issues/5406, 
https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615/2

It is true that without switching to eval(), after seeing large number of training images, the network will produce lower "validation" error, but this is not a real validation, because BatchNorm will be using stats from those training images, which it does not have access to in a real validation setting. 
To see really would happen if you don't switch to eval() mode, comment the line which switches to eval() mode and run test_model.py (not the main.py!). test_model.py loads a trained model and runs it on a bunch of images. Without eval(), you will be getting very bad accuracy numbers (0.5 last time I tried, which is the same as coinflip). 

Hope this helps.

**Comment by wowotou1022** [comment]
Thanks for your reply.
I used the images of the training set to do the test, and the results also got worse. This proved to be not a problem of generalization of the model.
For example, when testing "combine", the value of "bitwise-error" during "train mode" is 0.1609, and the value of "test mode" is 0.2268.
Can I test with "validate-trained-models.py"?

**Comment by ando-khachatryan** [comment]
Validation error is always slightly higher than training error. Even be present if you "validate" on the training set itself this effect may still be there, as I said, this is due to different behavior of BatchNorm in training and eval modes. 

However, validation error is too large when using the "combined noise", I can confirm this from my experiments as well. My understand is that there is something wrong, but I cannot tell what the reason is. I think I know how to investigate the issue but unfortunately not enough time at the moment.

Yes, run validate-trained-models.py for validation, sorry for a mistake in my previous post (mentioning test_model.py).

**Comment by wowotou1022** [comment]
Okay, thanks for your help!
